src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.AnalyzerExecutionContext.cs

```diff
@@ -139,7 +139,10 @@ HostSymbolStartAnalysisScope getSymbolAnalysisScopeCore()
                             lock (_gate)
```

Comment:
>lock (_gate) [](start = 28, length = 12)

I didn't understand what was the race condition. Could you share some of your analysis? Thanks

Reply:
Ah, sorry, I should have added more description. My original implementation was assuming that `getSymbolAnalysisScopeCore` can only be executed once per-symbol per-analyzer, `_lazySymbolScopeTasks` handling at start of `GetSymbolAnalysisScopeAsync` above will ensure that. However, in presence of cancellation getting fired while executing symbol start actions, we clear `_lazySymbolScopeTasks` and attempt to re-execute the symbol start actions (see [here](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/AnalyzerManager.cs,99)). In this scenario, we will end up with duplicate attempts at executing this code, and we will hit key already added to `_lazyPendingMemberSymbolsMapOpt` scenario.

---
In reply to: [223858883](https://github.com/dotnet/roslyn/pull/30399#discussion_r223858883) [](ancestors = 223858883)